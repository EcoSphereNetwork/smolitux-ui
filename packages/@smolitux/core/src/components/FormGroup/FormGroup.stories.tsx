import type { Meta, StoryObj } from '@storybook/react';
import React from 'react';
import { FormGroup } from './FormGroup';

const meta: Meta<typeof FormGroup> = {
  title: 'Core/FormGroup',
  component: FormGroup,
  parameters: { layout: 'centered' },
  tags: ['autodocs'],
  argTypes: {
    label: { control: 'text' },
    hint: { control: 'text' },
    error: { control: 'text' },
    required: { control: 'boolean' },
    horizontal: { control: 'boolean' },
    description: { control: 'text' },
  },
};
export default meta;

type Story = StoryObj<typeof FormGroup>;

export const Default: Story = {
  args: {
    label: 'Username',
    children: <input aria-label="Username" />,
  },
};

export const WithHint: Story = {
  args: {
    label: 'Email',
    hint: 'We will not share it',
    children: <input aria-label="Email" />,
  },
};

export const WithError: Story = {
  args: {
    label: 'Password',
    error: 'Too weak',
    required: true,
    children: <input aria-label="Password" />,
  },
};

export const Horizontal: Story = {
  args: {
    label: 'City',
    horizontal: true,
    labelFor: 'city',
    children: <input id="city" aria-label="City" />,
  },
};
