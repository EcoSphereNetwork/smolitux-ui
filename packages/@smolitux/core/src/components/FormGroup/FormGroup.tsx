import React, { forwardRef, useId } from 'react';
import clsx from 'clsx';

export interface FormGroupProps extends React.HTMLAttributes<HTMLFieldSetElement | HTMLDivElement> {
  /** Visible label for the group */
  label: string;
  /** Form elements to group */
  children: React.ReactNode;
  /** Additional descriptive text */
  description?: string;
  /** Optional hint text */
  hint?: string;
  /** Error message */
  error?: string;
  /** Mark group as required */
  required?: boolean;
  /** Display label and field horizontally */
  horizontal?: boolean;
  /** id of the form control the label describes */
  labelFor?: string;
  /** Position of the label relative to children */
  labelPosition?: 'top' | 'left';
  /** Id attribute */
  id?: string;
}

export const FormGroup = forwardRef<HTMLFieldSetElement | HTMLDivElement, FormGroupProps>(
  (
    {
      label,
      children,
      description,
      hint,
      error,
      required = false,
      horizontal = false,
      labelFor,
      labelPosition = 'top',
      id,
      className,
      ...rest
    },
    ref,
  ) => {
    const autoId = useId();
    const groupId = id || `form-group-${autoId}`;
    const hintId = hint ? `${groupId}-hint` : undefined;
    const errorId = error ? `${groupId}-error` : undefined;
    const descId = description ? `${groupId}-desc` : undefined;
    const describedBy = [errorId, hintId, descId].filter(Boolean).join(' ') || undefined;

    const Root = labelFor ? 'div' : 'fieldset';
    const Label = labelFor ? 'label' : 'legend';

    return (
      <Root
        ref={ref as any}
        id={groupId}
        role={labelFor ? 'group' : undefined}
        aria-labelledby={`${groupId}-label`}
        aria-required={required || undefined}
        aria-invalid={!!error || undefined}
        aria-describedby={describedBy}
        className={clsx('smx-form-group', horizontal && 'flex items-start gap-2', className)}
        {...rest}
      >
        <Label
          id={`${groupId}-label`}
          htmlFor={labelFor}
          className={clsx(
            labelPosition === 'left' && !horizontal ? 'mr-2' : 'mb-1',
            'font-medium'
          )}
        >
          {label}
          {required && (
            <span className="text-red-500 ml-1" aria-hidden="true">
              *
            </span>
          )}
        </Label>
        <div className={clsx(labelPosition === 'left' && horizontal && 'flex-1')}>{children}</div>
        {description && (
          <p id={descId} className="text-sm text-gray-500 mt-1">
            {description}
          </p>
        )}
        {hint && (
          <p id={hintId} className="text-sm text-gray-500 mt-1">
            {hint}
          </p>
        )}
        {error && (
          <p id={errorId} role="alert" className="text-sm text-red-600 mt-1">
            {error}
          </p>
        )}
      </Root>
    );
  },
);

FormGroup.displayName = 'FormGroup';
export default FormGroup;
