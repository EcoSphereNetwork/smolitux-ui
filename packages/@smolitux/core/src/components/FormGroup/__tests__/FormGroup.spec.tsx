import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { FormGroup } from '../FormGroup';

describe('FormGroup', () => {
  test('snapshot', () => {
    const tree = renderer
      .create(
        <FormGroup label="Name" id="fg">
          <input id="name" />
        </FormGroup>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders hint and error', () => {
    render(
      <FormGroup label="Email" hint="enter" error="invalid" id="fg">
        <input id="email" />
      </FormGroup>
    );
    expect(screen.getByText('enter')).toBeInTheDocument();
    expect(screen.getByRole('alert')).toHaveTextContent('invalid');
  });

  test('label focuses input', async () => {
    const user = userEvent.setup();
    render(
      <FormGroup label="Age" labelFor="age" id="fg">
        <input id="age" aria-label="Age" />
      </FormGroup>
    );
    await user.click(screen.getByText('Age'));
    expect(screen.getByLabelText('Age', { selector: 'input' })).toHaveFocus();
  });

  test('has correct aria attributes', () => {
    render(
      <FormGroup label="City" required error="e" hint="h" id="fg" labelFor="city">
        <input id="city" aria-label="City" />
      </FormGroup>
    );
    const group = screen.getByRole('group');
    expect(group).toHaveAttribute('aria-required', 'true');
    expect(group).toHaveAttribute('aria-invalid', 'true');
    const described = group.getAttribute('aria-describedby');
    expect(described).toContain('fg-error');
    expect(described).toContain('fg-hint');
  });
});
