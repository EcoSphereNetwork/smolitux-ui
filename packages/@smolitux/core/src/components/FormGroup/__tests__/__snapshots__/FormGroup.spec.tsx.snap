// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormGroup snapshot 1`] = `
<fieldset
  aria-labelledby="fg-label"
  className="smx-form-group"
  id="fg"
>
  <legend
    className="mb-1 font-medium"
    id="fg-label"
  >
    Name
  </legend>
  <div
    className=""
  >
    <input
      id="name"
    />
  </div>
</fieldset>
`;
